pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,351 @@
+
+		oids := []string{"1.3.6.1.2.1.1.2.0"}
+		v, err := params.Get(oids)
+		if err != nil || len(v.Variables) < 1 || v.Variables[0].Value == nil {
+			log.Debugf("SNMP get to %s error: %v", deviceIP, err)
+			l.deleteService(entityID)
+		} else {
+			log.Debugf("SNMP get to %s success: %v", deviceIP, v.Variables[0].Value)
+			l.createService(deviceIP, adIdentifier, entityID)
+		}
+	}
+}
+
+func (l *SNMPListener) checkDevices() {
+	subnets := []snmpSubnet{}
+	for i, config := range l.config.Configs {
+		ipAddr, ipNet, err := net.ParseCIDR(config.Network)
+		if err != nil {
+			return
+		}
+
+		defaultParams, err := config.BuildSNMPParams()
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		adIdentifier := fmt.Sprintf("snmp_%d", i)
+
+		startingIP := ipAddr.Mask(ipNet.Mask)
+		currentIP := make(net.IP, len(startingIP))
+		copy(currentIP, startingIP)
+
+		configHash := config.Digest(config.Network)
+		cacheKey := fmt.Sprintf("snmp:%s", configHash)
+		l.loadCache(config, adIdentifier, cacheKey)
+
+		subnet := snmpSubnet{
+			adIdentifier:  adIdentifier,
+			config:        config,
+			defaultParams: defaultParams,
+			startingIP:    startingIP,
+			currentIP:     currentIP,
+			network:       *ipNet,
+			cacheKey:      cacheKey,
+		}
+		subnets = append(subnets, subnet)
+	}
+
+	if l.config.Workers == 0 {
+		l.config.Workers = 2
+	}
+
+	if l.config.DiscoveryInterval == 0 {
+		l.config.DiscoveryInterval = 3600
+	}
+
+	jobs := make(chan snmpSubnet)
+	for w := 0; w < l.config.Workers; w++ {
+		go worker(l, jobs)
+	}
+
+	subnetsDone := 0
+	started := time.Now()
+	discoveryInterval := time.Duration(l.config.DiscoveryInterval) * time.Second
+
+	for {
+		for _, subnet := range subnets {
+			if subnet.network.Contains(subnet.currentIP) {
+				ignored := false
+				currentString := subnet.currentIP.String()
+				for _, ip := range subnet.config.IgnoredIPAddresses {
+					if ip == currentString {
+						ignored = true
+						break
+					}
+				}
+				if ignored {
+					incrementIP(subnet.currentIP)
+					continue
+				}
+				jobSubnet := subnet
+				jobSubnet.currentIP = make(net.IP, len(subnet.currentIP))
+				copy(jobSubnet.currentIP, subnet.currentIP)
+				jobs <- jobSubnet
+				incrementIP(subnet.currentIP)
+			} else {
+				// XXX write the cache a bit more often
+				writeCache(subnet.cacheKey, l.devices[subnet.adIdentifier])
+				subnetsDone++
+			}
+		}
+
+		var delay time.Duration = 0
+		if subnetsDone == len(subnets) {
+			for _, subnet := range subnets {
+				copy(subnet.currentIP, subnet.startingIP)
+			}
+			subnetsDone = 0
+			current := time.Now()
+			elapsed := current.Sub(started)
+			delay = discoveryInterval - elapsed
+		}
+
+		select {
+		case <-l.stop:
+			return
+		case <-time.After(delay):
+			if delay > 0 {
+				started = time.Now()
+			}
+		}
+	}
+}
+
+func (l *SNMPListener) createService(deviceIP string, adIdentifier string, entityID string) {
+	_, present := l.services[entityID]
+	if present {
+		return
+	}
```

Comment:
```suggestion
	if _, present := l.services[entityID]; present {
		return
	}
```
should we log something?